hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingContextUnorderedP.java

```diff
@@ -0,0 +1,374 @@
+                    .append(entry(broadcastKey(Keys.LAST_EMITTED_WM), lastReceivedWm))
+                    .onFirstNull(() -> snapshotTraverser = null);
+        }
+        return emitFromTraverserToSnapshot(snapshotTraverser);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected void restoreFromSnapshot(@Nonnull Object key, @Nonnull Object value) {
+        if (key instanceof BroadcastKey) {
+            assert ((BroadcastKey) key).key().equals(Keys.LAST_EMITTED_WM) : "Unexpected key: " + key;
+            // we restart at the oldest WM any instance was at at the time of snapshot
+            minRestoredWm = Math.min(minRestoredWm, (long) value);
+            return;
+        }
+        Tuple2<T, Integer> value1 = (Tuple2<T, Integer>) value;
+        // we can't apply backpressure here, we have to store the items and execute them later
+        for (int i = 0; i < value1.f1(); i++) {
+            restoredObjects.add(value1.f0());
+            LoggingUtil.logFinest(getLogger(), "Restored: %s", value1.f0());
+        }
+    }
+
+    @Override
+    public boolean finishSnapshotRestore() {
+        for (T t; (t = restoredObjects.peek()) != null && processItem(t); ) {
+            restoredObjects.remove();
+        }
+        if (restoredObjects.isEmpty()) {
+            // finish current object, we can't return true with a half-emitted item
+            if (!emitFromTraverser(currentTraverser)) {
+                return false;
+            }
+            restoredObjects = new ArrayDeque<>(0); // minimize the internal storage
+            lastReceivedWm = minRestoredWm;
+            logFine(getLogger(), "restored lastReceivedWm=%s", minRestoredWm);
+            return true;
+        } else {
+            tryFlushQueue();
+        }
+        return false;
+    }
+
+    @Override
+    public void close() {
+        // close() might be called even if init() was not called.
+        // Only destroy the context if is not shared (i.e. it is our own).
+        if (contextObject != null && !contextFactory.isSharedLocally()) {
+            contextFactory.destroyFn().accept(contextObject);
+        }
+        contextObject = null;
+    }
+
+    /**
+     * Drains items from the queue until either:
+     * <ul><li>
+     *     encountering an incomplete item
+     * </li><li>
+     *     the outbox gets full
+     * </li></ul>
+     *
+     * @return true if there are no more in-flight items and everything was emitted
+     *         to the outbox
+     */
+    @SuppressWarnings("unchecked")
+    private boolean tryFlushQueue() {
+        for (;;) {
+            if (!emitFromTraverser(currentTraverser)) {
+                return false;
+            }
+            Tuple3<T, Long, Object> tuple = resultQueue.poll();
+            if (tuple == null) {
+                return watermarkCounts.isEmpty();
+            }
+            asyncOpsCounter--;
+            inFlightItems.merge(tuple.f0(), -1, (o, n) -> o == 1 ? null : o - 1);
```

Comment:
It never does. The -1 would only be used if the old value wasn't present to go to -1. It could go below zero if there were multiple responses for single request, but this should not happen with CompletableFuture. I added an assert.